### ISSUE TITLE
Get Stack Overflow with OAuth

### ISSUE THREAD

**Issue by SledgeHammer01** [yidongnan/grpc-spring-boot-starter#457]
Hi,

I am using:

Spring Boot 2.4.0
grpc-server-spring-boot-starter 2.10.1.RELEASE
spring-security-oauth2-autoconfigure 2.2.9.RELEASE    (for OAuth token endpoint)

Beans are configured as:

```
	@Bean
	public GrpcAuthenticationReader grpcAuthenticationReader() {
	    return new BearerAuthenticationReader(token -> new BearerTokenAuthenticationToken(token));
	}
	
	@Bean
    public ProtobufJsonFormatHttpMessageConverter protobufHttpMessageConverter() {
        return new ProtobufJsonFormatHttpMessageConverter();
    }	

```

With no security enabled, my gRPC endpoint is working as expected.

With security enabled, I get a stack overflow exception:

It looks like in ProviderManager : 201, in the authenticate method, the BearerTokenAuthenticationToken calls the authenticate method itself.

```txt
Exception in thread "grpc-default-executor-0" java.lang.StackOverflowError
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
	at com.sun.proxy.$Proxy291.authenticate(Unknown Source)
	at org.springframework.security.authentication.ProviderManager.authenticate(ProviderManager.java:201)
	at org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter$AuthenticationManagerDelegator.authenticate(WebSecurityConfigurerAdapter.java:510)
	at jdk.internal.reflect.GeneratedMethodAccessor223.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
	at com.sun.proxy.$Proxy291.authenticate(Unknown Source)
	at org.springframework.security.authentication.ProviderManager.authenticate(ProviderManager.java:201)
	at org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter$AuthenticationManagerDelegator.authenticate(WebSecurityConfigurerAdapter.java:510)
```

**Comment by ST-DDT** [comment]
Could you please also include the grpc related stack lines?
Please debug which AuthenticationProviders are configured in that ProviderManager.
From what I can tell there is only a web `AuthenticationManagerDelegator` that calls itself. If there are other beans in there, try removing the looped one. If there are no other beans in there, then the oauth library doesnt setup the ProviderManager correctly/at all and probably uses their own OAuthAuthenticationProvider(?) bean. Check for it in your application context or in the configured web servlet filters and setup a ProviderManager bean that delegates to that bean.

**Comment by SledgeHammer01** [comment]
> Could you please also include the grpc related stack lines?
> Please debug which AuthenticationProviders are configured in that ProviderManager.
> From what I can tell there is only a web `AuthenticationManagerDelegator` that calls itself. If there are other beans in there, try removing the looped one. If there are no other beans in there, then the oauth library doesnt setup the ProviderManager correctly/at all and probably uses their own OAuthAuthenticationProvider(?) bean. Check for it in your application context or in the configured web servlet filters and setup a ProviderManager bean that delegates to that bean.

@ST-DDT 

Hi,

Thanks for the response. Here is the call stack when it enters ProviderManager::authenticate.

When I step through authenticate, it looks like this.providers only has AnonymousAuthenticationProvider. When I step through the regular HTTP Oauth authentication in swagger (please note, I am trying to do both HTTP and GRPC in the same app as shown in the Medium sample) it has AnonymousAuthenticationProvider and DaoAuthenticationProvider.

It skips over the provider since it fails on the supports() call.

Then it gets to this part:

		if (result == null && this.parent != null) {
			// Allow the parent to try.
			try {
				parentResult = this.parent.authenticate(authentication);
				result = parentResult;
			}

parent is pointing to a JdkDynamicAopProxy for org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter$AuthenticationManagerDelegator@2119bb3d

authentication = BearerTokenAuthenticationToken

this call: parentResult = this.parent.authenticate(authentication); just keeps calling authenticate recursively.

When I try get a token through Swagger for the HTTP side (which is working), the behavior is different.

authentication = UsernamePasswordAuthenticationToken and the DaoAuthenticationProvider handles the request.

ProviderManager.authenticate(Authentication) line: 167	
WebSecurityConfigurerAdapter$AuthenticationManagerDelegator.authenticate(Authentication) line: 518	
DefaultAuthenticatingServerInterceptor.interceptCall(ServerCall<ReqT,RespT>, Metadata, ServerCallHandler<ReqT,RespT>) line: 103	
ServerInterceptors$InterceptCallHandler<ReqT,RespT>.startCall(ServerCall<ReqT,RespT>, Metadata) line: 244	
ExceptionTranslatingServerInterceptor.interceptCall(ServerCall<ReqT,RespT>, Metadata, ServerCallHandler<ReqT,RespT>) line: 61	
ServerInterceptors$InterceptCallHandler<ReqT,RespT>.startCall(ServerCall<ReqT,RespT>, Metadata) line: 244	
MetricCollectingServerInterceptor.interceptCall(ServerCall<Q,A>, Metadata, ServerCallHandler<Q,A>) line: 138	
ServerInterceptors$InterceptCallHandler<ReqT,RespT>.startCall(ServerCall<ReqT,RespT>, Metadata) line: 244	
Contexts.interceptCall(Context, ServerCall<ReqT,RespT>, Metadata, ServerCallHandler<ReqT,RespT>) line: 52	
GrpcRequestScope.interceptCall(ServerCall<ReqT,RespT>, Metadata, ServerCallHandler<ReqT,RespT>) line: 75	
ServerInterceptors$InterceptCallHandler<ReqT,RespT>.startCall(ServerCall<ReqT,RespT>, Metadata) line: 244	
ServerImpl$ServerTransportListenerImpl.startWrappedCall(String, ServerMethodDefinition<WReqT,WRespT>, ServerStream, Metadata, CancellableContext, Tag) line: 648	
ServerImpl$ServerTransportListenerImpl.startCall(ServerStream, String, ServerMethodDefinition<ReqT,RespT>, Metadata, CancellableContext, StatsTraceContext, Tag) line: 626	
ServerImpl$ServerTransportListenerImpl.access$1900(ServerImpl$ServerTransportListenerImpl, ServerStream, String, ServerMethodDefinition, Metadata, Context$CancellableContext, StatsTraceContext, Tag) line: 416	
ServerImpl$ServerTransportListenerImpl$1StreamCreated.runInternal() line: 557	
ServerImpl$ServerTransportListenerImpl$1StreamCreated.runInContext() line: 532	
ServerImpl$ServerTransportListenerImpl$1StreamCreated(ContextRunnable).run() line: 37	
SerializingExecutor.run() line: 123	
ThreadPoolExecutor.runWorker(ThreadPoolExecutor$Worker) line: 1128	
ThreadPoolExecutor$Worker.run() line: 628	
Thread.run() line: 834

**Comment by SledgeHammer01** [comment]
> Could you please also include the grpc related stack lines?
> Please debug which AuthenticationProviders are configured in that ProviderManager.
> From what I can tell there is only a web `AuthenticationManagerDelegator` that calls itself. If there are other beans in there, try removing the looped one. If there are no other beans in there, then the oauth library doesnt setup the ProviderManager correctly/at all and probably uses their own OAuthAuthenticationProvider(?) bean. Check for it in your application context or in the configured web servlet filters and setup a ProviderManager bean that delegates to that bean.

@ST-DDT 

I've been debugging how the working HTTP oauth works and it goes through:

```
public class BearerTokenExtractor implements TokenExtractor {

	private final static Log logger = LogFactory.getLog(BearerTokenExtractor.class);

	@Override
	public Authentication extract(HttpServletRequest request) {
		String tokenValue = extractToken(request);
		if (tokenValue != null) {
			PreAuthenticatedAuthenticationToken authentication = new PreAuthenticatedAuthenticationToken(tokenValue, "");
			return authentication;
		}
		return null;
	}

```

I tried using PreAuthenticatedAuthenticationToken, but it still loops. Seems like the working Oauth path goes through the Spring OAuth2AuthenticationProcessingFilter and OAuth2AuthenticationManager whereas GPRC doesn't hit that at all.

**Comment by ST-DDT** [comment]
> I tried using PreAuthenticatedAuthenticationToken, but it still loops. Seems like the working Oauth path goes through the Spring OAuth2AuthenticationProcessingFilter and OAuth2AuthenticationManager whereas GPRC doesn't hit that at all.

I'm not sure whether I picked the right oauth source code:

https://github.com/spring-projects/spring-security-oauth/blob/ef8d78403d551ca9c23c6de4db0aa6aae2e24d79/spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/authentication/OAuth2AuthenticationManager.java#L42

````java
public class OAuth2AuthenticationManager implements AuthenticationManager, ...
````

https://github.com/yidongnan/grpc-spring-boot-starter/blob/28d304db5144f1ef8ce7933649f0f47bc91b008f/grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/security/interceptors/DefaultAuthenticatingServerInterceptor.java#L71-L72

As you can see that class already implements the right interface, but for some reasons there seems to be a different (primary) `AuthenticationManager` bean, so all we have to do is the following:

````java
@Bean
public DefaultAuthenticatingServerInterceptor authenticatingServerInterceptor(
            OAuth2AuthenticationManager authenticationManager,
            GrpcAuthenticationReader authenticationReader) {
        return new DefaultAuthenticatingServerInterceptor(authenticationManager, authenticationReader);
}


@Bean
GrpcAuthenticationReader oauthAuthReader() {
    return new BearerToken...(token -> new PreAuthenticatedAuthenticationToken(tokenValue, ""));
}
````
(or somehow get rid of the other `AuthenticationManager` bean, or redefine it the oauth one as primary, or configure the primary one to include the OAuth one...)

If you wish to support both Username/Password (Basic) auth and oAuth for grpc, then you can do that be using the `CompositeGrpcAuthenticationReader`

> Whereas with the gRPC, it's going into authenticate() on every call. That might be the issue.

Due to the structure of grpc (it doesn't care about spring-securities context holder), I cannot update the security context once the actual authorization check is required, so I have to eagerly authenticate the token.

See also https://github.com/yidongnan/grpc-spring-boot-starter/blob/master/grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/security/interceptors/DefaultAuthenticatingServerInterceptor.java#L109-L126

**Comment by SledgeHammer01** [comment]
> > I tried using PreAuthenticatedAuthenticationToken, but it still loops. Seems like the working Oauth path goes through the Spring OAuth2AuthenticationProcessingFilter and OAuth2AuthenticationManager whereas GPRC doesn't hit that at all.
> 
> I'm not sure whether I picked the right oauth source code:
> 
> https://github.com/spring-projects/spring-security-oauth/blob/ef8d78403d551ca9c23c6de4db0aa6aae2e24d79/spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/authentication/OAuth2AuthenticationManager.java#L42
> 
> ```java
> class OAuth2AuthenticationManager implements AuthenticationManager, ...
> ```
> 
> https://github.com/yidongnan/grpc-spring-boot-starter/blob/28d304db5144f1ef8ce7933649f0f47bc91b008f/grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/security/interceptors/DefaultAuthenticatingServerInterceptor.java#L71-L72
> 
> As you can see that class already implements the right interface, but for some reasons there seems to be a different (primary) `AuthenticationManager` bean, so all we have to do is the following:
> 
> ```java
> @Bean
> DefaultAuthenticatingServerInterceptor authenticatingServerInterceptor(
>             OAuth2AuthenticationManager authenticationManager,
>             GrpcAuthenticationReader authenticationReader) {
>         return new DefaultAuthenticatingServerInterceptor(authenticationManager, authenticationReader);
> }
> 
> 
> @Bean
> GrpcAuthenticationReader oauthAuthReader() {
>     return new BearerToken...(token -> new PreAuthenticatedAuthenticationToken(tokenValue, ""));
> }
> ```
> 
> (or somehow get rid of the other `AuthenticationManager` bean, or redefine the oauth one as primary, or configure the primary one to include the OAuth one...)
> 
> If you wish to support both Username/Password (Basic) auth and oAuth for grpc, then you can do that be using the `CompositeGrpcAuthenticationReader`
> 
> > Whereas with the gRPC, it's going into authenticate() on every call. That might be the issue.
> 
> Due to the structure of grpc (it doesn't care about spring-securities context holder), I cannot update the security context once the actual authorization check is required, so I have to eagerly authenticate the token.
> 
> See also https://github.com/yidongnan/grpc-spring-boot-starter/blob/master/grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/security/interceptors/DefaultAuthenticatingServerInterceptor.java#L109-L126

Yes, that is the correct source :). The OAuth2AuthenticationManager doesn't seem like its exposed as a bean. Bean not found. It is created by @EnableResourceServer later in: ResourceServerSecurityConfigurer.configure(HttpSecurity) line: 207 when it sets up the OAuth2AuthenticationProcessingFilter. Just new'ed up. Not a bean. Not sure how I can get a ref to it. Seems like the ACTUAL AuthenticationManager bean is the WebSecurity one. The OAuth2AuthenticationManager just seems to be private to the filter.

Any ideas?

**Comment by ST-DDT** [comment]
> Any ideas?

Reflection?

Isn't [OAuth2AuthenticationManager](https://github.com/spring-projects/spring-security-oauth/blob/ef8d78403d551ca9c23c6de4db0aa6aae2e24d79/spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/authentication/OAuth2AuthenticationManager.java#L41) deprecated?
Is there some new spring security based replacement?

**Comment by SledgeHammer01** [comment]
> > Any ideas?
> 
> Reflection?
> 
> Isn't [OAuth2AuthenticationManager](https://github.com/spring-projects/spring-security-oauth/blob/ef8d78403d551ca9c23c6de4db0aa6aae2e24d79/spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/authentication/OAuth2AuthenticationManager.java#L41) deprecated?
> Is there some new spring security based replacement?

Spring used to have an OAuth2 authorization server component (the one I am using) and they decided to deprecate it for some reason with no replacement and the community was mad :). So they started a new project: Spring Authorization Server, but it is still in incubation. 

More testing and debugging confirms what I said above, that the "real" AuthenticationManager instance doesn't have the OAuth2AuthenticationManager in it since that's handled by the OAuth2 filter which has its own internal instance outside of everything else. I don't think there is any easy way to get to that instance (I was also thinking reflection).
 
I did found a solution though. I created a secondary instance of OAuth2AuthenticationManager just for grpc and it looks to be working. All it needs is the TokenServices bean to point to the database.

Looks like all is working now.

Thanks again for all the help!
